### PR TITLE
Rework and expand Pointing Device support

### PR DESCRIPTION
**Pull request by drashna**
## Description

Give pointing device support some much needed love. 

This adds a "driver" value for pointing device feature, so you can quickly and simply add support for a sensor to act as pointing device by specifying the sensor type. 
Additionally, it adds more robust user callbacks to help with these changes.
It also generalizes the sensor code, so that they don't have pointing device specific code, and can be used for other purposes (such as with HID joystick). 

## Types of Changes

- [x] Core
- [x] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/sensors/pimoroni_trackball.c [hunk 1]
```diff
@@ -17,73 +17,42 @@
...
 }
 
-i2c_status_t read_pimoroni_trackball(pimoroni_data* data) {
-    i2c_status_t status = i2c_readReg(PIMORONI_TRACKBALL_ADDRESS << 1, TRACKBALL_REG_LEFT, (uint8_t*)data, sizeof(*data), TRACKBALL_TIMEOUT);
-#ifdef TRACKBALL_DEBUG
-    dprintf("Trackball READ i2c_status_t: %d\nLeft: %d\nRight: %d\nUp: %d\nDown: %d\nSwtich: %d\n", status, data->left, data->right, data->up, data->down, data->click);
-#endif
+i2c_status_t read_pimoroni_trackball(pimoroni_data_t* data) {
+    i2c_status_t status = i2c_readReg(PIMORONI_TRACKBALL_ADDRESS << 1, PIMORONI_TRACKBALL_REG_LEFT, (uint8_t*)data, sizeof(*data), PIMORONI_TRACKBALL_TIMEOUT);
+    if (debug_mouse) dprintf("Trackball READ i2c_status_t: %d\nLeft: %d\nRight: %d\nUp: %d\nDown: %d\nSwtich: %d\n", status, data->left, data->right, data->up, data->down, data->click);
```
[the remark this answers is not in the document]
Comment (drashna) [reply drivers/sensors/pimoroni_trackball.c:45]: well, it's behind the debug_mouse setting already

[the remark this answers is not in the document]
Comment (daskygit) [reply drivers/sensors/pimoroni_trackball.c:45]: Okie dokies, it was just that at the default interval setting this outputs to the console 125 times a second. You wont be able to see anything from `mousekey_debug` which also uses `debug_mouse` to enable console logging.

What about the suggestion above? It just compacts the output a little.

[the remark this answers is not in the document]
Comment (drashna) [reply drivers/sensors/pimoroni_trackball.c:45]: Fair.  And I can add some sort of timing gate for the reports, so it's not spamming it too much.

**Comment by drashna** [comment]
@daskygit added timeout check for debug printing, so it shouldn't be too bad now.  And preprocessor statements to remove the code, if console isn't enabled.

**Comment by drashna** [comment]
@atulloh wanted to update the oddball to include the changes here.  If you wouldn't mind confirming they don't break anything?  I don't have a working adns9800 sensor ATM, so I can't verify, but the pmw3360 sensor does appear to work properly.

**Comment by atulloh** [comment]
> @atulloh wanted to update the oddball to include the changes here. If you wouldn't mind confirming they don't break anything? I don't have a working adns9800 sensor ATM, so I can't verify, but the pmw3360 sensor does appear to work properly.

@drashna Sure, I'll dig out that boards over the weekend and see if it's all good. Nice to see all this consolidated 👍

**Comment by drashna** [comment]
> > @atulloh wanted to update the oddball to include the changes here. If you wouldn't mind confirming they don't break anything? I don't have a working adns9800 sensor ATM, so I can't verify, but the pmw3360 sensor does appear to work properly.
> 
> @drashna Sure, I'll dig out that boards over the weekend and see if it's all good. Nice to see all this consolidated 👍

Awesome! 

And yeah, more boards are getting sensors, so it's been on my mind for a while now. 

Wanted to make it dead easy to add support for people, since otherwise it will be a lot of copy-paste that gets fragmented, etc.

**Comment by drashna** [comment]
@atulloh Have you had a chance to test this out?

**Comment by tzarc** [comment]
Conflicts need to be resolved.

File: drivers/sensors/adns9800.c [hunk 2]
```diff
@@ -15,83 +15,82 @@
...
 #define CPI_STEP 200
 #define CLAMP_CPI(value) value<MIN_CPI ? MIN_CPI : value> MAX_CPI ? MAX_CPI : value
-#define SPI_MODE 3
-#define SPI_DIVISOR (F_CPU / ADNS_CLOCK_SPEED)
 #define US_BETWEEN_WRITES 120
 #define US_BETWEEN_READS 20
 #define US_BEFORE_MOTION 100
 #define MSB1 0x80
 
 extern const uint8_t firmware_data[];
```
Comment (daskygit) [inline drivers/sensors/adns9800.c:80]:
```suggestion

```
This isn't needed any more right?

File: quantum/pointing_device.c [hunk 3]
```diff
@@ -1,34 +1,56 @@
...
+#endif
+#if (defined(POINTING_DEVICE_ROTATION_90) + defined(POINTING_DEVICE_ROTATION_180) + defined(POINTING_DEVICE_ROTATION_270)) > 1
+#    error More than one rotation selected.  This is not supported.
+#endif
 
 static report_mouse_t mouseReport = {};
 
+extern const pointing_device_driver_t pointing_device_driver;
+
 __attribute__((weak)) bool has_mouse_report_changed(report_mouse_t new, report_mouse_t old) { return (new.buttons != old.buttons) || (new.x&& new.x != old.x) || (new.y&& new.y != old.y) || (new.h&& new.h != old.h) || (new.v&& new.v != old.v); }
```
Comment (tzarc) [inline quantum/pointing_device.c:31]: This looks like it's incorrectly formatted?
Might also be worth looking at using `memcmp` as well, probably faster.

File: quantum/pointing_device_drivers.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
+/* Copyright 2017 Joshua Broekhuijsen <snipeye+qmk@gmail.com>
```
Comment (tzarc) [inline quantum/pointing_device_drivers.c:1]: Ideally these implementations would be in the driver code directly. Worth cleaning up later.

**Review by tzarc — CHANGES_REQUESTED** [review]
Nothing major.

File: quantum/pointing_device_drivers.c (hunk 4 above)
Comment (drashna) [reply quantum/pointing_device_drivers.c:1]: Yeah, I had thought about that, and considered it, but wanted to stick with a more standard reporting for each sensor, as some support buttons and some don't.   And I didn't really want to introduce a new report just for the sensors.

File: quantum/pointing_device.c (hunk 3 above)
Comment (drashna) [reply quantum/pointing_device.c:31]: ... that's how lint likes it formatted ... 

But yeah, memcmp would probably be better here, in general

File: quantum/pointing_device_drivers.c (hunk 4 above)
Comment (drashna) [reply quantum/pointing_device_drivers.c:1]: also, I planned on looking at adding support to the joystick stuff, later on.  And will probably clean it up more before then.

**Review by tzarc — APPROVED** [review]

**Pull request merged**